lib/std/fs/Dir.zig

```diff
@@ -780,6 +780,137 @@ pub fn walk(self: Dir, allocator: Allocator) Allocator.Error!Walker {
+                if (top.iter.next() catch |err| {
+                    // If we get an error, then we want the user to be able to continue
+                    // walking if they want, which means that we need to pop the directory
+                    // that errored from the stack. Otherwise, all future `next` calls would
+                    // likely just fail with the same error.
+                    var item = self.stack.pop().?;
+                    if (self.stack.items.len != 0) {
+                        item.iter.dir.close();
+                    }
+                    return err;
+                }) |entry| {
+                    self.name_buffer.shrinkRetainingCapacity(dirname_len);
+                    if (self.name_buffer.items.len != 0) {
+                        try self.name_buffer.append(allocator, fs.path.sep);
+                        dirname_len += 1;
+                    }
+                    try self.name_buffer.ensureUnusedCapacity(allocator, entry.name.len + 1);
+                    self.name_buffer.appendSliceAssumeCapacity(entry.name);
+                    self.name_buffer.appendAssumeCapacity(0);
+                    if (mode == .automatic and entry.kind == .directory) {
+                        try self.enter(allocator, entry);
+                    }
+                    return .{
+                        .dir = top.iter.dir,
+                        .basename = self.name_buffer.items[dirname_len .. self.name_buffer.items.len - 1 :0],
+                        .path = self.name_buffer.items[0 .. self.name_buffer.items.len - 1 :0],
+                        .kind = entry.kind,
+                    };
+                } else {
+                    var item = self.stack.pop().?;
+                    if (self.stack.items.len != 0) {
+                        item.iter.dir.close();
+                    }
+                }
+            }
+            return null;
+        }
+
+        /// Traverses into the directory, continuing walking one level down.
+        pub fn enter(self: *Self, allocator: Allocator, entry: Self.Entry) !void {
+            if (entry.kind != .directory) {
+                @branchHint(.cold);
+                return;
+            }
+
+            var new_dir = entry.dir.openDir(entry.basename, .{ .iterate = true }) catch |err| {
+                switch (err) {
+                    error.NameTooLong => unreachable,
+                    else => |e| return e,
+                }
+            };
+
+            errdefer new_dir.close();
+
+            try self.stack.append(allocator, .{
+                .iter = new_dir.iterateAssumeFirstIteration(),
+                .dirname_len = self.name_buffer.items.len - 1,
+            });
+        }
+
+        pub fn deinit(self: *Self, allocator: Allocator) void {
+            self.name_buffer.deinit(allocator);
+            self.stack.deinit(allocator);
+        }
+
+        pub fn depth(self: *Self) usize {
+            return self.stack.items.len;
+        }
```

Comment:
I don't think this implementation works for the hybrid version, since `self.stack.items.len` will be artificially increased for `directory` entries in automatic mode, e.g.:

```
dir/
├─ a/
│  └─ ...
└─ b
```

In manual mode `depth()` will return `1` when `entry.basename` is `a`, but in automatic mode it'll erroneously return `2` (and when `entry.basename` is `b`, both modes will return `1`).

Reply:
hum... maybe there's no good way of knowing the depth in automatic mode without keeping track of the last value we yielded, which might be bad. It served mostly as a debug for me and perhaps we can live without it?